### PR TITLE
r.surf.gauss: Added seed option and flag

### PR DESCRIPTION
**Pull request by cwhite911**
Addresses issue https://github.com/OSGeo/grass/issues/1044 to add seed parameter and flag for reproducible results to the r.surf.gauss module. However, tests are currently not present and should still be included.

### CODE REVIEW

**Comment by wenzeslaus** [comment]
I know this is a draft, but just to be clear, the compilation fails for `raster/r.surf.gauss`.

File: raster/r.surf.gauss/main.c [hunk 1]
```diff
@@ -55,9 +58,43 @@ int main(int argc, char *argv[])
...
+        G_math_srand(seed_value);
+        G_verbose_message(_("Read random seed from %s option: %ld"), seed->key,
+                          seed_value);
+    }
+    else {
+        /* default as it used to be */
+        seed_value = G_math_srand_auto();
+        G_verbose_message(_("Warning set flag s or option seed. Generated "
+                            "random seed (-s): %ld"),
+                          seed_value);
```
Comment (petrasovaa) [inline raster/r.surf.gauss/main.c:95]: This should probably be G_warning, no?

**Review by petrasovaa — CHANGES_REQUESTED** [review]
It looks like the test is not testing anything except that the tool runs.

File: raster/r.surf.gauss/main.c (hunk 1 above)
Comment (wenzeslaus) [reply raster/r.surf.gauss/main.c:95]: Speaking from from strict API point of view, it should be message, later warning, and when possible to break compatibly an error. I would have to investigate to see what we did in the past (if message or warning).

**Comment by cwhite911** [comment]
@petrasovaa I updated the tests to check the univar stats are correct,

**Comment by nilason** [comment]
The Labeler went nearly all-in for this. @echoix Do you perhaps know why?

**Comment by echoix** [comment]
I think it's because it was an old PR, and merging main in it meant a lot of changes in everywhere. But I'm not sure. Now we don't remove labels automatically, so a next commit should only use the ones that are defined.

If it is still unexpected, maybe a rebase would do the trick...

**Comment by nilason** [comment]
> ...merging main in it meant a lot of changes in everywhere....

That sounds plausible.

**Comment by echoix** [comment]
> > ...merging main in it meant a lot of changes in everywhere....
> 
> 
> 
> That sounds plausible.

It doesn't worry that me that much.

**Comment by petrasovaa** [comment]
@cwhite911 see the discussion we had in https://github.com/OSGeo/grass/pull/3480#discussion_r1519351514, we basically decided not to use the -s flag, so this PR needs to be updated...

File: raster/r.surf.gauss/main.c [hunk 2]
```diff
@@ -55,9 +58,43 @@ int main(int argc, char *argv[])
...
         exit(EXIT_FAILURE);
 
+    /****** INITIALISE RANDOM NUMBER GENERATOR ******/
+    if (s_flag->answer) {
+        seed_value = G_math_srand_auto();
+        G_verbose_message(_("Generated random seed (-s): %ld"), seed_value);
+    }
+    else if (seed->answer) {
+        seed_value = atol(seed->answer);
+        G_math_srand(seed_value);
```
Comment (petrasovaa) [inline raster/r.surf.gauss/main.c:86]: Use `G_srand48` and ` G_srand48_auto` instead.

**Comment by cwhite911** [comment]
> > > ...merging main in it meant a lot of changes in everywhere....
> > 
> > 
> > That sounds plausible.
> 
> It doesn't worry that me that much.

This might of been my fault. When was catching this project up I rebased and had to go back and fix the branch by resetting it to your commit and cherry-picking the work I just did.

File: raster/r.surf.gauss/main.c [hunk 3]
```diff
@@ -55,9 +57,32 @@ int main(int argc, char *argv[])
...
     sigma->answer = "1.0";
 
+    seed = G_define_option();
+    seed->key = "seed";
+    seed->type = TYPE_INTEGER;
+    seed->required = NO;
+    seed->label = _("Seed for random number generator");
+    seed->description = _("The same seed can be used to obtain same results"
+                          " or random seed can be generated by other means.");
+
```
Comment (echoix) [inline raster/r.surf.gauss/main.c:67]: Do we have standardized messages for common options between modules?

File: raster/r.surf.gauss/main.c [hunk 4]
```diff
@@ -55,9 +57,32 @@ int main(int argc, char *argv[])
...
+        seed_value = atol(seed->answer);
+        G_srand48(seed_value);
+        G_verbose_message(_("Read random seed from %s option: %ld"), seed->key,
+                          seed_value);
+    }
+    else {
+        /* default as it used to be */
+        seed_value = G_math_srand_auto();
+        G_warning(_("Warning set option seed. Generated "
+                    "random seed: %ld"),
```
Comment (echoix) [inline raster/r.surf.gauss/main.c:82]: I have trouble understanding that sentence. I assume the word "Warning" is superfluous, but the remainder doesn't seem right neither. I think I already wrote a comment on this specific sentence somewhere else.

File: raster/r.surf.gauss/testsuite/test_surf_gauss.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+
+"""
+MODULE:    Test of r.surf.gauss
+
+AUTHOR(S): Corey White <ctwhite48 gmail com>
+
+PURPOSE: Tests random gauss surface generation
+
+COPYRIGHT: (C) 2023 by Corey White and the GRASS Development Team
```
Comment (echoix) [inline raster/r.surf.gauss/testsuite/test_surf_gauss.py:10]:
```suggestion
COPYRIGHT: (C) 2024 by Corey White and the GRASS Development Team
```

Wasn't merged in 2023. Or if it was written in 2023, add a date range

File: raster/r.surf.gauss/main.c (hunk 3 above)
Comment (petrasovaa) [reply raster/r.surf.gauss/main.c:67]: Not really, we would have to add a new standard option (something like G_OPT_M_SEED) to parser. That would make sense, but maybe as a separate PR?

File: raster/r.surf.gauss/testsuite/test_surf_gauss.py (hunk 5 above)
Comment (cwhite911) [reply raster/r.surf.gauss/testsuite/test_surf_gauss.py]: I set the date range to 2023  - 2024.

File: raster/r.surf.gauss/main.c (hunk 4 above)
Comment (cwhite911) [reply raster/r.surf.gauss/main.c]: I rewrote the warning to: 

```c
G_warning(_("Autogenerated random seed set to: %ld"), seed_value);
```

File: raster/r.surf.gauss/main.c [hunk 6]
```diff
@@ -55,9 +57,30 @@ int main(int argc, char *argv[])
...
+    if (seed->answer) {
+        seed_value = atol(seed->answer);
+        G_srand48(seed_value);
+        G_verbose_message(_("Read random seed from %s option: %ld"), seed->key,
+                          seed_value);
+    }
+    else {
+        /* default as it used to be */
+        seed_value = G_math_srand_auto();
+        G_warning(_("Autogenerated random seed set to: %ld"), seed_value);
```
Comment (petrasovaa) [inline raster/r.surf.gauss/main.c:81]:
```suggestion
        G_verbose_message(_("Autogenerated random seed set to: %ld"), seed_value);
```

File: raster/r.surf.gauss/testsuite/test_surf_gauss.py
Comment (petrasovaa) [inline raster/r.surf.gauss/testsuite/test_surf_gauss.py:1]: Naming convention seems to be `test_r_surf_gauss.py`

File: raster/r.surf.gauss/testsuite/test_surf_gauss.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+    @classmethod
+    def setUpClass(cls):
+        """Ensures expected computational region"""
+        os.environ["GRASS_RANDOM_SEED"] = "42"
+        # modifying region just for this script
+        cls.use_temp_region()
+        # Only 100,000,000 seem to resonably (not 100%) ensure that all values
+        # are generated, so exceeding of ranges actually shows up.
+        cls.runModule("g.region", rows=10000, cols=10000)
```
Comment (petrasovaa) [inline raster/r.surf.gauss/testsuite/test_surf_gauss.py:36]: The test runs 46 s for me, that's unnecessarily long...

File: raster/r.surf.gauss/testsuite/test_surf_gauss.py
Comment (cwhite911) [reply raster/r.surf.gauss/testsuite/test_surf_gauss.py:1]: Good catch I didn't notice that.

File: raster/r.surf.gauss/testsuite/test_surf_gauss.py (hunk 7 above)
Comment (cwhite911) [reply raster/r.surf.gauss/testsuite/test_surf_gauss.py]: I reduced the extent to 1000X1000 and it now runs in about a second.

Comment (echoix) [reply raster/r.surf.gauss/testsuite/test_surf_gauss.py]: What's the smallest size that tests the same things? Is 5x12 or other enough?

Comment (cwhite911) [reply raster/r.surf.gauss/testsuite/test_surf_gauss.py]: I'm not sure what the appropriate spatial scale is to capture all of the variability in the Gaussian random field. I did some basic tests that fail if we go lower than 1000x1000 unless we lower the precision when checking the univariate stats.

The other option is to just test if the seed options are getting set correctly for this PR.

Comment (petrasovaa) [reply raster/r.surf.gauss/testsuite/test_surf_gauss.py]: I changed the tests to just check that the r.univar values match. The seed variable was anyway set to 42, so the results were always the same even if the seed option was not used.

**Review by petrasovaa — APPROVED** [review]

**Pull request merged**